app/reconciliation.go

```diff
@@ -0,0 +1,148 @@
+package app
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/cosmos-sdk/store/iavl"
+)
+
+const reconStoreKey = "reconciliation"
+
+// unbalancedBlockHeightKey for saving unbalanced block height for reconciliation
+var unbalancedBlockHeightKey = []byte("0x01")
+
+var (
+	SupplyKey          = banktypes.SupplyKey
+	DenomAddressPrefix = banktypes.DenomAddressPrefix
+	BalancesPrefix     = banktypes.BalancesPrefix
+)
+
+// reconBalance will do reconciliation for accounts balances.
+func (app *App) reconBalance(ctx sdk.Context, bankIavl *iavl.Store) {
+	if ctx.BlockHeight() <= 2 {
+		return
+	}
+
+	height, exists := app.getUnbalancedBlockHeight(ctx)
+	if exists {
+		panic(fmt.Sprintf("unbalanced state at block height %d, please use hardfork to bypass it", height))
+	}
+
+	balanced := app.getBankChanges(ctx, bankIavl)
+	bankIavl.ResetDiff()
+
+	if !balanced {
+		app.saveUnbalancedBlockHeight(ctx)
+	}
+}
+
+func (app *App) getBankChanges(ctx sdk.Context, bankIavl *iavl.Store) bool {
+	supplyPre := sdk.Coins{}
+	balancePre := sdk.Coins{}
+	supplyCurrent := sdk.Coins{}
+	balanceCurrent := sdk.Coins{}
+
+	diff := bankIavl.GetDiff()
+	version := ctx.BlockHeight() - 2
+	ctx.Logger().Debug("reconciliation changes", "height", ctx.BlockHeight(), "version", version)
```

Comment:
Also please research does Iavl have the `prune all` strategy, so that if a node enable such strategy, it can not sync.

Reply:
When using pruning = "everything", there are at least two states. When using custom pruning strategy, it also needs at least two states. Then it should be fine.